bench/wtperf/wtperf.c

```diff
@@ -131,6 +146,173 @@ randomize_value(CONFIG_THREAD *thread, char *value_buf)
+	    thread->workload->truncate_count * trunc->truncation_percentage;
+	/* How many milestones we need */
+	trunc->needed_milestones = thread->workload->truncate_count /
+	    trunc->truncate_milestone_gap;
+
+	final_milestone_gap = trunc->truncate_milestone_gap;
+
+	/* Reset this value for use again */
+	trunc->truncate_milestone_gap = 0;
+
+	/*
+	 * Here we check if there is data in the collection. If there is
+	 * data available, then we need to setup some initial truncation
+	 * milestones.
+	 */
+	ret = cursor->next(cursor);
+
+	if (ret == 0) {
+		ret = cursor->get_key(cursor, &key);
+	}
+	/* We have data */
+	if (ret == 0) {
+		start_point_val = decode_key(cfg, key);
+		cursor->reset(cursor);
+		if ((ret = cursor->prev(cursor)) != 0) {
+			lprintf(cfg, ret, 0,
+			    "truncate setup end: prev failed");
+			return (-1);
+		}
+		if ((ret = cursor->get_key(cursor,
+			    &key)) != 0) {
+			lprintf(cfg, ret, 0,
+			    "truncate setup end: get_key failed");
+			return (-1);
+		}
+		end_point_val = decode_key(cfg, key);
+		/* Not enough documents? */
+			if (start_point_val + trunc->needed_milestones
+			    > end_point_val)
+				trunc->truncate_milestone_gap = 0;
+			else
+				trunc->truncate_milestone_gap =
+				    (end_point_val - start_point_val)
+				    / trunc->needed_milestones;
+	} else if ( ret != WT_NOTFOUND ) {
+		lprintf(cfg, ret, 0,
+			    "truncate setup start: get_key failed");
+		printf("ret was %d\n", ret);
+		return (-1);
+	}
+
+	/* If we have enough data alloc some milestones */
+	if (trunc->truncate_milestone_gap != 0) {
+		trunc->expected_total = (end_point_val - start_point_val);
+		for (i = 0; i < trunc->needed_milestones; i++) {
+			truncate_key = calloc(cfg->key_sz, 1);
+			truncate_item =
+			    calloc(sizeof(TRUNCATE_QUEUE_ENTRY), 1);
+			if (truncate_item == NULL) {
+				(void)enomem(cfg);
+				return (-1);
+			}
+			generate_key(cfg, truncate_key,
+			    trunc->truncate_milestone_gap * (i+1));
+			truncate_item->key = truncate_key;
+			truncate_item->diff =
+			    (trunc->truncate_milestone_gap * (i+1)) -
+			    trunc->last_key;
+			STAILQ_INSERT_TAIL( &cfg->truncate_stone_head,
+			    truncate_item, q);
+			trunc->last_key = trunc->truncate_milestone_gap * (i+1);
+			trunc->num_milestones++;
+		}
+	}
+	trunc->truncate_milestone_gap = final_milestone_gap;
+	return (0);
+}
+
+static int
+run_truncate(CONFIG *cfg, CONFIG_THREAD *thread,
+    TRUNCATE_CONFIG *trunc, WT_CURSOR *cursor, WT_SESSION *session) {
+
+	char *truncate_key;
+	int ret;
+	size_t i;
+	TRUNCATE_QUEUE_ENTRY *truncate_item;
+
+	/* Update the total inserts */
+	trunc->total_gross_inserts = 0;
+	for (i=0; i < thread->cfg->workers_cnt; i++) {
+		trunc->total_gross_inserts +=
+		    thread->cfg->workers[i].total_inserts;
+	}
+
+	trunc->expected_total +=
+	    (trunc->total_gross_inserts - trunc->last_total_inserts);
+	trunc->last_total_inserts = trunc->total_gross_inserts;
+
+	/* We have enough data, so we can look to truncate */
```

Comment:
This code is setting up the milestones, not doing the truncate. I find the comment confusing.